### PR TITLE
[ready to review] SIFT descriptor as in https://github.com/ducha-aiki/pytorch-sift

### PR DESCRIPTION
**Pull request by ducha-aiki**
-  [x] implementation
-  [x] docs
-  [x] tests

### CODE REVIEW

**Comment by ducha-aiki** [comment]
I have added a flag to output even-sized Gaussian and refactored the Gaussian filter itself.
No change for odd input, but proper symmetrical for even. Why do we even need even-sized Gaussian? To able to calculate SIFT of even sizes, like 32x32, 64x64, etc.

File: kornia/feature/siftdesc.py [hunk 1]
```diff
@@ -1,12 +1,14 @@
 from typing import Tuple
-import kornia
 import torch
 import math
 import torch.nn as nn
 import torch.nn.functional as F
-import numpy as np
+from kornia.filters import get_gaussian_kernel2d
+from kornia.filters import spatial_gradient
+from math import pi
```
Comment (edgarriba) [inline kornia/feature/siftdesc.py]: we have kornia.pi which already cast pi number into a torch tensor. We can discuss the use of it.

File: kornia/feature/siftdesc.py [hunk 2]
```diff
@@ -18,154 +20,139 @@ def get_sift_pooling_kernel(ksize:int = 25)-> torch.Tensor:
...
-        self.gx.weight.data = torch.tensor([[[[-1.0, 0.0, 1.0]]]]).float()
-        
-        self.gy = nn.Conv2d(1, 1, kernel_size=(3,1),  bias = False)
-        self.gy.weight.data = torch.tensor([[[[-1], [0], [1]]]]).float()
-        
-        nw = get_sift_pooling_kernel(ksize = self.bin_ksize).float()
-        
+
+        ks: int = self.patch_size
+        sigma: float = float(ks) / math.sqrt(2.0)
```
Comment (edgarriba) [inline kornia/feature/siftdesc.py]: maybe is a bit nitpick, but we can try to use pure torch for basic stuff ? just for convention with the library, no more arguments for it.

File: kornia/feature/siftdesc.py [hunk 3]
```diff
@@ -18,154 +20,139 @@ def get_sift_pooling_kernel(ksize:int = 25)-> torch.Tensor:
...
-        for x in range(ksize):
-            disq = (y - ksize_2+0.5)**2 +  (x - ksize_2+0.5)**2;
-            kernel[y,x] = math.exp(-disq / sigma_mul_2)
-            if circ and (disq >= r2):
-                kernel[y,x] = 0.
-    return kernel
 
 class SIFTDescriptor(nn.Module):
     """
     Module, which computes SIFT descriptors of given patches
```
Comment (edgarriba) [inline kornia/feature/siftdesc.py]: I would add documentation for the arguments and maybe a small snippet example.

File: kornia/filters/kernels.py [hunk 4]
```diff
@@ -5,12 +5,10 @@
 
 
 def gaussian(window_size, sigma):
```
Comment (edgarriba) [inline kornia/filters/kernels.py]: add typing

File: test/feature/test_siftdesc.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+def test_get_sift_bin_ksize_stride_pad(ps,
+                                       n_bins,
+                                       ksize,
+                                       stride,
+                                       pad):
+    out = get_sift_bin_ksize_stride_pad(ps, n_bins)
+    assert out == (ksize, stride, pad)
+
+
+class TestSIFTDescriptor:
```
Comment (edgarriba) [inline test/feature/test_siftdesc.py]: what happens with patches with more than one channel ?

Comment (ducha-aiki) [reply test/feature/test_siftdesc.py]: Crash and it is intended: original SIFT is for single channels. Stuff like CSIFT is a less standard. 
Although I can add the flag to just apply SIFT to all channels, so the output would be multichannel as well. What do you think?

File: kornia/feature/siftdesc.py (hunk 1 above)
Comment (ducha-aiki) [reply kornia/feature/siftdesc.py]: fixed

File: test/feature/test_siftdesc.py (hunk 5 above)
Comment (edgarriba) [reply test/feature/test_siftdesc.py]: looks good, go for it

**Pull request merged**